File: java/L100/LeetCode0188.java
package L100;

public class LeetCode0188 {
    public int maxProfit(int k, int[] prices) {
        if (prices == null || prices.length < 2) {
            return 0;
        }
        final int N = prices.length;
        int[][][] profits = new int[N][N / 2 + 1][2];
        profits[0][0][0] = 0 - prices[0];
        for (int i = 1;i < N / 2 + 1;++i) {
            profits[0][i][0] = Integer.MIN_VALUE / 2;
            profits[0][i][1] = Integer.MIN_VALUE / 2;
        }
        for (int i = 1;i < N;++i) {
            profits[i][0][0] = Math.max(profits[i - 1][0][0], profits[i - 1][0][1] - prices[i]);
            for (int j = 1;j < N / 2 + 1;++j) {
                profits[i][j][0] = Math.max(profits[i - 1][j][1] - prices[i], profits[i - 1][j][0]);
                profits[i][j][1] = Math.max(profits[i - 1][j - 1][0] + prices[i], profits[i - 1][j][1]);
            }
        }
        int ans = 0;
        k = Math.min(k, N / 2);
        for (int i = 0;i <= k;++i) {
            ans = Math.max(ans, profits[N - 1][i][1]);
        }
        return ans;
    }
}

File: cpp/L100/LeetCode0188.cc
#include <limits>
#include <vector>
using std::vector;

class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        vector<vector<int>> dp(k + 1, vector<int>(2, 0));
        for (int i = 0;i <= k;++i) {
            dp[i][1] = std::numeric_limits<int>::min();
        }
        for (int price : prices) {
            for (int i = k;i >= 1;--i) {
                int t = dp[i][0];
                dp[i][0] = std::max(dp[i][0], dp[i - 1][1] + price);
                dp[i][1] = std::max(dp[i][1], t - price);
            }
            dp[0][1] = std::max(dp[0][1], -price);
        }
        int profit{0};
        for (int i = 0;i <= k;++i) {
            profit = std::max(profit, std::max(dp[i][0], dp[i][1]));
        }
        return profit;
    }
};